<div id="top"></div>
<!--



-->
[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]



<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/Gero1999/code/new/main/R/DEA">
    <img src="DNA_microarray.svg" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">Differential Expression Analysis</h3>

  <p align="center">
    Analyze multiple datasets without even sweating!
    <br />
    <br />
    <br />
    <a href="https://github.com/Gero1999/code/tree/main/R/DEA">View Demo</a>
    Â·
  </p>
</div>



<!-- ABOUT THE PROJECT -->
## Aims of the project


The purpose of this project is to be capable of automatizing pipeline processes in order to:

* Select in a faster way from all the microarray repositories the datasets of your interest
* Simplify metadata wihtout the intervention of the user
* Elaborate an exploratory and differential expression analysis comparing two or more situations.
* Use genes significantly expressed in order to perform a pathway analysis.

<p align="right">(<a href="#top">back to top</a>)</p>


## What is a differential expression analysis?

Differential expression analysis is a fundamental technique used in bioinformatics and genomics to identify genes or transcripts that exhibit significant changes in expression levels between different conditions or experimental groups. The primary goal of this analysis is to determine which genes are upregulated or downregulated under different experimental conditions, such as different treatments, disease states, or developmental stages.

The process of differential expression analysis typically involves comparing gene expression levels in two or more sample groups using high-throughput sequencing data, such as RNA-seq. First, the raw sequencing data is preprocessed and aligned to a reference genome or transcriptome. Next, gene expression levels are quantified, often as read counts or fragments per kilobase of transcript per million mapped reads (FPKM). Statistical methods, such as the negative binomial model or the edgeR method, are then applied to identify genes that show significant differences in expression between the sample groups.

The results of differential expression analysis provide valuable insights into the molecular mechanisms underlying different biological processes, helping researchers understand how gene expression patterns change in response to specific treatments or conditions. This analysis is widely used in genomics, transcriptomics, and other omics studies to gain a deeper understanding of complex biological processes and to identify potential biomarkers and therapeutic targets.

<p align="right">(<a href="#top">back to top</a>)</p>


### Built With R

* [Biobase]()
* [ArrayExpress]()
* [Hugene]()
* [Oligo]()
* [Limma]()
* [clusterProfiler]()
* [ggplot2]()
* [dplyr]()
* [stringr]()
* [arrayQualityMetrics]()
* [geneplotter]()
* [enrichR]()
* [vsn]()



<p align="right">(<a href="#top">back to top</a>)</p>



<!-- STILL IN DEVELOP! -->




<!-- USAGE EXAMPLES -->
## Still in develop!




