
<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/Gero1999/code/new/main/shinyR/pairwise_alignment">
    <img src="dsDNA_linear.png" alt="Logo" width="80" height="80">
  </a>

  <h3 align="center">Pair-Aligner</h3>
  <a href="https://gero1999.shinyapps.io/Pair-Aligner/"><strong>Check the app Â»</strong></a>
  <p align="center">
    See how your sequences align! 
  </p>
</div>



<!-- ABOUT THE PROJECT -->
## About The Project


What are the uses of this app? 
* Compare the homology of two sequences graphically 
* Obtain an score to assess an alignment
* Use dynamic programming (Watermann algorithm) to deduce the optimal solution

<p align="right">(<a href="#top">back to top</a>)</p>


## What is the Waterman algorithm?

The Waterman algorithm, also known as the Smith-Waterman algorithm, is an optimal solution for local sequence alignment in bioinformatics. It is specifically designed to identify the regions of similarity between two sequences, rather than aligning the entire sequences. This algorithm uses dynamic programming to efficiently construct a matrix that represents all possible alignments between the sequences. Each cell of the matrix stores the score of the best alignment ending at that position, and the score is determined based on a scoring scheme for matches, mismatches, and gaps.

As a local alignment tool, the Waterman algorithm allows researchers to pinpoint the most similar regions between the sequences, which is crucial for tasks like identifying conserved regions or functional domains in DNA or protein sequences.

In summary, the Waterman algorithm is an efficient and widely-used method for local sequence alignment, providing researchers with valuable insights into the regions of similarity between sequences.




<!-- GETTING STARTED -->
## 


